### PR TITLE
api: add listener_address_restriction_hint in Node to hint known listening ports on the node

### PR DESCRIPTION
**Pull request by sanjaypujare**
Description: Added an optional field in `Node` as a generic hint to the management server about known listening ports on the node

Risk Level: Low
Testing: manually tested by compiling protos into generated Java code and using the code in a Java project
Docs Changes: N/A
Release Notes: N/A

  Signed-off-by: Sanjay Pujare <sanjaypujare@users.noreply.github.com>

### CODE REVIEW

**Comment by sanjaypujare** [comment]
@htuch pls review

**Comment by ejona86** [comment]
To be more clear about the use-case: This is for cases where the server has already decided which ports it will listen on and needs configuration for those ports.

**Comment by sanjaypujare** [comment]
Looked into the `envoy-linux (format)` failure:
```
-import "envoy/config/core/v3/base.proto";
+import "envoy/api/v3/core/socket_option.proto";
```

For some reason `socket_option.proto` was not properly placed in the v3 hierarchy.  It should go into `envoy/config/core/v3`. I think I need to update some file to include socket_option.proto there.

**Comment by alyssawilk** [comment]
@htuch @lizan given this is API work one of you up for taking this?

**Comment by markdroth** [comment]
This was suggested by @htuch, so I assume he'll review.

**Comment by sanjaypujare** [comment]
@htuch Just FYI I did try to test the changes on my Mac but there are infra/tooling problems I am unable to debug. Pls let me know if the failures are related to my changes.

File: api/envoy/api/v2/core/base.proto [hunk 1]
```diff
@@ -182,6 +184,12 @@ message Node {
   // See :ref:`the list of features <client_features>` that xDS client may
   // support.
   repeated string client_features = 10;
+
+  // Known listening ports on the node as a generic hint to the management server
+  // for filtering :ref:`listeners <config_listeners>` to be returned. For example,
+  // if there is a listener bound to port 80, the list can optionally contain the
+  // SocketAddress `(0.0.0.0,80)`. The field is optional and just a hint.
+  repeated Address listener_address_restriction_hint = 11;
```
Comment (htuch) [inline api/envoy/api/v2/core/base.proto]: This PR looks good, just trying to see if we can maybe name this a bit more descriptively (I know I was responsible for the original suggestion :P)

How about `listener_bound_addresses`, `listening_addresses`, `listener_available_addresses`?

Comment (sanjaypujare) [reply api/envoy/api/v2/core/base.proto]: I'll pick `listening_addresses` as the shortest of them all.

**Comment by sanjaypujare** [comment]
> LGTM, but you need to run `fix_format` (to forward propagate into v3) ...

So I ran 
```
./ci/run_envoy_docker.sh './ci/do_ci.sh fix_format'
```
And that finished successfully and generated following changes:
```
	modified:   api/BUILD
	modified:   api/envoy/api/v2/core/base.proto
	modified:   api/envoy/api/v2/core/socket_option.proto
	modified:   api/envoy/api/v2/listener.proto
	modified:   api/envoy/config/bootstrap/v2/bootstrap.proto
	modified:   api/envoy/config/bootstrap/v3/BUILD
	modified:   api/envoy/config/bootstrap/v3/bootstrap.proto
	modified:   api/envoy/config/core/v3/BUILD
	modified:   api/envoy/config/core/v3/address.proto
	modified:   api/envoy/config/core/v3/base.proto
	modified:   api/envoy/config/listener/v3/BUILD
	modified:   api/envoy/config/listener/v3/listener.proto
	modified:   generated_api_shadow/envoy/api/v2/core/address.proto
	modified:   generated_api_shadow/envoy/api/v2/core/base.proto
	modified:   generated_api_shadow/envoy/api/v2/listener.proto
	modified:   generated_api_shadow/envoy/config/bootstrap/v2/bootstrap.proto
	modified:   generated_api_shadow/envoy/config/bootstrap/v3/BUILD
	modified:   generated_api_shadow/envoy/config/bootstrap/v3/bootstrap.proto
	modified:   generated_api_shadow/envoy/config/core/v3/BUILD
	modified:   generated_api_shadow/envoy/config/core/v3/address.proto
	modified:   generated_api_shadow/envoy/config/core/v3/base.proto
	modified:   generated_api_shadow/envoy/config/listener/v3/BUILD
	modified:   generated_api_shadow/envoy/config/listener/v3/listener.proto

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	api/envoy/api/v3/
	generated_api_shadow/envoy/api/v2/core/socket_option.proto
	generated_api_shadow/envoy/api/v3/

```

Should I commit all these changes? And more importantly, why is `socket_options`  under `api/envoy/api/v3/core/` instead of `api/envoy/config/core/v3` like other similar files?

File: api/envoy/api/v2/core/socket_option.proto [hunk 2]
```diff
@@ -0,0 +1,47 @@
+syntax = "proto3";
+
+package envoy.api.v2.core;
+
+import "validate/validate.proto";
```
Comment (htuch) [inline api/envoy/api/v2/core/socket_option.proto]: You'll need to add option `(udpa.annotations.file_migrate).move_to_package = "envoy.config.core.v3";` (as done in other files in this package) for the correct file movement in v3.
`

**Comment by sanjaypujare** [comment]
> @sanjaypujare good catch on the v3 file move:
> 
> 1. Yes, please add all the new shadow files.
> 2. See my comment on how to ensure the new socket options file is in the right place.
>    /wait

@htuch done. PTAL.

**Comment by htuch** [comment]
@sanjaypujare I think you ended up colliding with a known master breakage. Can you sync past https://github.com/envoyproxy/envoy/pull/10426, merge master and try again?

You will also need to fix DCO: https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#fixing-dco

**Comment by sanjaypujare** [comment]
> @sanjaypujare I think you ended up colliding with a known master breakage. Can you sync past #10426, merge master and try again?
> 
> You will also need to fix DCO: https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#fixing-dco

Done both. PTAL.

**Comment by htuch** [comment]
@sanjaypujare see the docs warning:
```
checking consistency... /source/generated/rst/api-v3/config/core/v3/socket_option.proto.rst: WARNING: document isn't included in any toctree
```
I suggest adding to https://github.com/envoyproxy/envoy/blob/master/docs/root/api-v3/common_messages/common_messages.rst  and https://github.com/envoyproxy/envoy/blob/master/docs/root/api-v2/common_messages/common_messages.rst as appropriate.

**Comment by sanjaypujare** [comment]
@htuch v3 common_messages.rst updated (v2 was already done).

File: api/envoy/api/v2/core/base.proto [hunk 3]
```diff
@@ -182,6 +185,12 @@ message Node {
   // See :ref:`the list of features <client_features>` that xDS client may
   // support.
   repeated string client_features = 10;
+
+  // Known listening ports on the node as a generic hint to the management server
+  // for filtering :ref:`listeners <config_listeners>` to be returned. For example,
+  // if there is a listener bound to port 80, the list can optionally contain the
+  // SocketAddress `(0.0.0.0,80)`. The field is optional and just a hint.
+  repeated Address listening_addresses = 11;
```
Comment (lizan) [inline api/envoy/api/v2/core/base.proto]: This is not implemented yet in this PR? so hide from doc? or do you plan to implement them?

Comment (sanjaypujare) [reply api/envoy/api/v2/core/base.proto]: Not sure about what is "not implemented". We are planning to use this new field in our xDS client (proxyless gRPC) as a hint to the xDS server for particular listener(s). I don't see any reason hide it in the docs. 

Also note that we had some offline discussion with @htuch about this and this was the agreed upon approach.

Comment (htuch) [reply api/envoy/api/v2/core/base.proto]: @lizan the user specifics it in bootstrap, and this then gets automagically dumped into each xDS stream (we merge `Node` messages).

Comment (lizan) [reply api/envoy/api/v2/core/base.proto]: @htuch ah ok I see how it works now. Thanks.

**Comment by sanjaypujare** [comment]
> LGTM

@htuch thanks! Once this is merged the downstream design/implementation activities will continue.

**Pull request merged**